### ISSUE TITLE
CentOS 7 reconfiguration failure

### ISSUE THREAD

**Issue by schwartz-je** [latchset/pkcs11-provider#337]
When recompiling from the master branch, we have to use 'autoreconf -fi'
This operation failed on a standard CentOS7 host
```
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
```
$ cat /etc/redhat-release 
CentOS Linux release 7.9.2009 (Core)
```
```
$ rpm -q gcc libtool m4 pkgconfig autoconf autoconf-archive automake
gcc-4.8.5-44.el7.x86_64
libtool-2.4.2-22.el7_3.x86_64
m4-1.4.16-10.el7.x86_64
pkgconfig-0.27.1-4.el7.x86_64
autoconf-2.69-11.el7.noarch
autoconf-archive-2017.03.21-1.el7.noarch
automake-1.13.4-3.el7.noarch
```
```
$ unzip pkcs11-provider-main-c9fb942.zip
   creating: pkcs11-provider-main/
  inflating: pkcs11-provider-main/.clang-format
...
  inflating: pkcs11-provider-main/tests/turi
finishing deferred symbolic links:
  pkcs11-provider-main/README -> README.md
$ cd pkcs11-provider-main
$ autoreconf -fi
aclocal: warning: couldn't open directory 'm4': Aucun fichier ou dossier de ce type
libtoolize: putting auxiliary files in `.'.
libtoolize: copying file `./ltmain.sh'
libtoolize: putting macros in `m4'.
libtoolize: copying file `m4/libtool.m4'
libtoolize: copying file `m4/ltoptions.m4'
libtoolize: copying file `m4/ltsugar.m4'
libtoolize: copying file `m4/ltversion.m4'
libtoolize: copying file `m4/lt~obsolete.m4'
libtoolize: Consider adding `AC_CONFIG_MACRO_DIR([m4])' to configure.ac and
libtoolize: rerunning libtoolize, to keep the correct libtool macros in-tree.
configure.ac:110: error: possibly undefined macro: PKG_CHECK_VAR
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1
```
To avoid this, I have to modify the 'configure.ac' file like this
```
--- configure.ac.orig   2023-09-12 14:55:12.000000000 +0200
+++ configure.ac        2023-10-06 14:32:25.768591147 +0200
@@ -8,8 +8,28 @@
 AM_INIT_AUTOMAKE([foreign dist-xz -Wall -Werror])
 AC_CONFIG_MACRO_DIRS([m4])

+## https://github.com/varnish/libvmod-example/pull/18/commits/ad96d3bc5ebe126c8cb6f03591c6fc62ffd6f3d6
+##
+# backwards compat with older pkg-config (< 0.28)
+# - pull in AC_DEFUN from pkg.m4
+m4_ifndef([PKG_CHECK_VAR], [
+# PKG_CHECK_VAR(VARIABLE, MODULE, CONFIG-VARIABLE,
+# [ACTION-IF-FOUND], [ACTION-IF-NOT-FOUND])
+# -------------------------------------------
+# Retrieves the value of the pkg-config variable for the given module.
+AC_DEFUN([PKG_CHECK_VAR],
+[AC_REQUIRE([PKG_PROG_PKG_CONFIG])dnl
+AC_ARG_VAR([$1], [value of $3 for $2, overriding pkg-config])dnl
+_PKG_CONFIG([$1], [variable="][$3]["], [$2])
+AS_VAR_COPY([$1], [pkg_cv_][$1])
+AS_VAR_IF([$1], [""], [$5], [$4])dnl
+])# PKG_CHECK_VAR
+])
+
 # Checks for programs.
 AC_PROG_CC
+### add for legacy reasons
+AM_PROG_CC_C_O
 AM_PROG_AR
 LT_INIT
 PKG_PROG_PKG_CONFIG
```

After patching, all is correct
```
$ autoreconf -fi
libtoolize: putting auxiliary files in `.'.
libtoolize: copying file `./ltmain.sh'
libtoolize: putting macros in `m4'.
libtoolize: copying file `m4/libtool.m4'
libtoolize: copying file `m4/ltoptions.m4'
libtoolize: copying file `m4/ltsugar.m4'
libtoolize: copying file `m4/ltversion.m4'
libtoolize: copying file `m4/lt~obsolete.m4'
libtoolize: Consider adding `AC_CONFIG_MACRO_DIR([m4])' to configure.ac and
libtoolize: rerunning libtoolize, to keep the correct libtool macros in-tree.
configure.ac:33: installing './ar-lib'
configure.ac:32: installing './compile'
configure.ac:34: installing './config.guess'
configure.ac:34: installing './config.sub'
configure.ac:8: installing './install-sh'
configure.ac:8: installing './missing'
src/Makefile.am: installing './depcomp'
parallel-tests: installing './test-driver'
```

**Comment by simo5** [comment]
pkcs11-provider only makes sense in combination with OpenSSL 3 which is not shipped in CentOS 7 (nor even CentOS 8).
I think you should keep it as a downstream patch.

**Comment by schwartz-je** [comment]
Hi Simo,
I develop the PKCS#11 interface for our ATOS Crypt2Pay HSM, and most of our client want keep their private keys into this HSM.

Following the recommendations of the french security institute (ANSSI),
some of them decided to use OpenSSL-3 instead of 1.1.1, and I need to help them to use the PKCS#11 provider instead of our PKCS#11 engine.
Among them, we can find many different kind of UNIX
- RHEL-7
- RHEL-8
- AIX
That's why I build your provider against CentOS7 and AIX 7.1 (an issue will soon arrive).

Regards,
Jean-Etienne

Le 29/01/2024 à 17:24, Simo Sorce a écrit :

Caution: External email. Do not open attachments or click links, unless this email comes from a known sender and you know the content is safe.


pkcs11-provider only makes sense in combination with OpenSSL 3 which is not shipped in CentOS 7 (nor even CentOS 8).
I think you should keep it as a downstream patch.

**Comment by simo5** [comment]
I see.

If your patches add backwards compatibility without breaking modern platforms I will accept pull requests for them.

But you will have to maintain them over the long term, we have no testing over old platforms (although if you know of a way of adding github actions that can test on them that is a possibility).

**Comment by space88man** [comment]
@schwartz-je - disclaimer untested:

Have you considered creating dist on a platform with OpenSSL 3
and using the tar.gz on your CentOS 7

```
# example on  el8 with openssl3-devel from epel (i.e. 
# an OS with no native OpenSSL 3
autoreconf -fi
CRYPTO_CFLAGS="-I/usr/include/openssl3" \
CRYPTO_LIBS="-L/usr/lib64/openssl3 -lcrypto"  \
OPENSSL_CFLAGS="-I/usr/include/openssl3" \
OPENSSL_LIBS="-L/usr/lib64/openssl3 -lssl -lcrypto" \
./configure

make dist

# now copy pkcs11-provider-0.3.tar.gz to 
# your CentOS 7 target with your ownbrew
# OpenSSL 3
```

**Comment by schwartz-je** [comment]
In fact, this patch is made to fix the use of  'autoreconf' on a CentOS7 (7.9.2009) host with standard development packages.
The main error is that the package 'pkgconfig' is just one release too old for the pkcs11-provider configuration.

OpenSSL 3.0 builds without any troubles.

Regards,
Jean-Etienne

Le 13/03/2024 à 01:14, space88man a écrit :

Caution: External email. Do not open attachments or click links, unless this email comes from a known sender and you know the content is safe.


@schwartz-je<https://github.com/schwartz-je> - disclaimer untested:

Have you considered creating dist on a platform with OpenSSL 3
and using the tar.gz on your CentOS 7

# example on  el8 with openssl3-devel from epel (i.e.
# an OS with no native OpenSSL 3
autoreconf -fi
CRYPTO_CFLAGS="-I/usr/include/openssl3" \
CRYPTO_LIBS="-L/usr/lib64/openssl3 -lcrypto"  \
OPENSSL_CFLAGS="-I/usr/include/openssl3" \
OPENSSL_LIBS="-L/usr/lib64/openssl3 -lssl -lcrypto" \
./configure

make dist

# now copy pkcs11-provider-0.3.tar.gz to
# your CentOS 7 target with your ownbrew
# OpenSSL 3

**Comment by simo5** [comment]
Closing this one as well,
given no contributions came in form of a PR
